### ISSUE TITLE
Adopt eslint-config-standard-with-typescript

### ISSUE THREAD

**Issue by mightyiam** [standard/standard#1007]
I've been working on [eslint-config-standard-with-typescript](https://www.npmjs.com/package/eslint-config-standard-with-typescript) alone. It would be nice to have it adopted by the standard org and have some cooperation on it.

**Comment by mightyiam** [comment]
👅

**Comment by LinusU** [comment]
Hey, sorry that no one has responded to you. I would be happy to help you transition the package and help maintain it as well 🙌

**Comment by mightyiam** [comment]
What do we do?

**Comment by LinusU** [comment]
Saw that you invited me, I'll try and move it here 🙌

**Comment by LinusU** [comment]
ah, I cannot do that, but I think that you can initiate a transfer, and then I can accept it on our side

**Comment by mightyiam** [comment]
"You don’t have the permission to create repositories on standard" when trying to transfer it to the standard organization.

**Comment by mightyiam** [comment]
Try adding me to the standard organization?

**Comment by mightyiam** [comment]
I'm totally up for this to be adopted by @standard.

**Comment by PascalAOMS** [comment]
Maybe it progresses a bit faster if I leave a comment here. /shrug

**Comment by mightyiam** [comment]
@PascalAOMS the comment is in response to the stale bot adding the `stale` label.

I am still actively maintaining this package. And I've just updated it with all the breaking changes from upstream https://github.com/typescript-eslint/typescript-eslint/.

**Comment by LinusU** [comment]
Sorry for the delay here, I think that @feross or someone else that is `owner` on the standard reop will need to fix this 😬 unfortunately I don't have the permissions to fix this

**Comment by mightyiam** [comment]
:trumpet:

**Comment by despian** [comment]
+1 would love to see this integrated into standard

**Comment by bjufre** [comment]
Any news on this?

**Comment by swansontec** [comment]
I tried using `eslint-config-standard-with-typescript`, but found that the rules are much stricter than I would have expected. For instance, it requires every function to declare its return type, even though Typescript is perfectly capable of inferring this. It also enforces the `number[]` array style over `Array<number>`, even thought the later is perfectly valid.

I understand that `eslint-config-standard-with-typescript` is derived from the `@typescript-eslint/eslint-plugin` recommended preset. Unfortunately, that preset has a rather aggressive "everything but the kitchen sink" approach to enabling linting rules.

The beautiful thing about Standard.js is that it focuses on finding actual bugs, and gets "out of the way" for everything else. Requiring `number[]` over `Array<number>` doesn't have any clear bug-fighting benefit, for example, which makes it very out-of-character for a Standard.js variant. I do want to see Typescript support enter the Standard.js ecosystem, but it needs to take the same conservative approach that Standard.js takes, only enabling rules where there is a clear benefit to doing so.

**Comment by mightyiam** [comment]
@swansontec, thank you for giving it a try.

I welcome participation. Let's discuss all of the issues that you care about in the project's own issue tracker.

**Comment by LinusU** [comment]
> It also enforces the number[] array style over Array<number>, even thought the later is perfectly valid.

I personally think that this is great. There are two ways of doing something, and the enforced style is the one I've seen most in the wild.

I don't think that Standard "focuses on finding actual bugs, and gets "out of the way" for everything else", e.g. we do enforce a lot of white-space things, indentation and no semicolons.

Enforcing style as well as helping finding bugs helps keeps the code base uniform and helps readability, especially when there are multiple maintainers working on the code. For me, this is one of the awesome things about Standard ☺️

**Comment by feross** [comment]
Sorry for the delay on this. Still catching up on issues.

@LinusU I'm happy to help transfer this package to the @standard organization as long as you're involved in the maintenance and direction of the rules, to help out @mightyiam. I'm not using Typescript in too many places yet so I won't be able to chime in with my patented Taste as a Service™️. If something's living in the @standard org then it should follow a similar ethos and vibe.

If it goes well, we might be able to build TypeScript support into the main `standard` binary by default so it just works on TypeScript projects too. That would be neat.

**Comment by mightyiam** [comment]
Thank you for getting around to address this, @feross.

I did my best to follow the ethos and vibe of Standard, while also attempting to make as many decisions as I could reasonably make.

My desire from the get go was for it to become a part of @standard and I accept whatever that would mean in regards to ownership and control.

Looking forward to cooperate.

**Comment by feross** [comment]
> I did my best to follow the ethos and vibe of Standard, while also attempting to make as many decisions as I could reasonably make.

Great! Looking forward to collaborating on this. Let's wait to hear what @LinusU says.

**Comment by LinusU** [comment]
> If it goes well, we might be able to build TypeScript support into the main standard binary by default so it just works on TypeScript projects too. That would be neat.

😍 😍 😍 

> @LinusU I'm happy to help transfer this package to the @standard organization as long as you're involved in the maintenance and direction of the rules, to help out @mightyiam.

I think it sounds good, I'm happy to be involved in the maintenance and direction of the rules 👍

**Comment by mightyiam** [comment]
Lovely.

Would we like to make some changes before transferring it to @standard?

I'm referring to anything that might be non-representative of @standard that we might like to change.

**Comment by feross** [comment]
We can just do the transfer now. I trust that you and @LinusU will figure things out as you develop it here. 

@mightyiam I just sent you an invite to the @standard organization. Can you try transferring the repo here now? I am unable to transfer your personal repo.

**Comment by LinusU** [comment]
> Would we like to make some changes before transferring it to @standard?

I've been using it for a long time and have never seen something that I've considered non-standard ☺️

**Comment by mightyiam** [comment]
Transferred.

I remember something about this org having a policy of all issues in this particular repo. Is that still the case, please?

**Comment by feross** [comment]
I think that until we're ready to support Typescript in `standard` (if we decide it's a good idea) then it would be better to keep the issues separate for now. I think it will be easier to see all Typescript related issues in `eslint-config-standard-with-typescript` to avoid confusion for now.

**Comment by swansontec** [comment]
@LinusU @mightyiam Thanks, I'll go ahead and create issues in that repo, then.

Since the number of "extra rules" in `eslint-config-standard-with-typescript` surprised me, I actually created my own competing project, [eslint-config-standard-kit](https://www.npmjs.com/package/eslint-config-standard-kit), which creates a one-stop solution for mixing & matching Standard.js with Flow, Typescript, JSX, and / or Prettier.

Since it looks like `eslint-config-standard-with-typescript` is becoming the semi-official Standard.js + Typescript solution going forward, I would like to collaborate on bringing `eslint-config-standard-kit` in sync with its rules, especially the auto-fixable ones. Here is [my issue](https://github.com/swansontec/eslint-config-standard-kit/issues/1) for tracking the necessary changes.

**Comment by feross** [comment]
@swansontec The work you've done to separate all the rules into separate packs is excellent! I'd love to see how we could adopt that directly here eventually. In the meantime, I want to get a few more releases out. But this would be great to offer to users right out of the box!

**Comment by ndom91** [comment]
+1 would love "rule packs" like that, but out of the box typescript support is much higher on mine (and I'm assuming many others) lists :)

**Comment by jandppw** [comment]
We are jumping on this bandwagon. Is there anything we can do to make it progress?

**Comment by feross** [comment]
To support TS directly in `standard`, do we need to just add some config to `eslintrc.json` in this repo?

```js
  "overrides": {
    "files": ["*.ts", "*.tsx"],
    "extends": ["standard-with-typescript", "standard-jsx"]
  }
```

I think we might also need changes in `standard-engine`?

cc @mightyiam @toddbluhm

**Comment by Divlo** [comment]
@feross To support TypeScript, we should look to integrate `ts-standard` into the `standardjs` organization, that would be awesome!

**Comment by feross** [comment]
@Divlo Happy to do that if @toddbluhm thinks it will help more people see it/find it. But I'm more interested in offering out-of-the-box support for `.ts` files since it appears that this is built in to ESLint now.

**Comment by Divlo** [comment]
Oh well, out-of-the-box support for `.ts` files (I suppose for `.tsx` files too), that's an interesting solution.
What need to be done to support TypeScript out-of-the-box with `standard` ? I'm glad to help if needed.

**Comment by toddbluhm** [comment]
While I do like `ts-standard`, I completely agree with @feross and would rather have ts support built into `standard` itself. 

One issue to overcome that I had to figure out with `ts-standard` was setting/finding the location of the `tsconfig.json` file in each project. We can probably go with the assumption that the `tsconfig.json` file is always in the root folder and then implement a flag to allow passing a custom path to the file. The `tsconfig.json` file is required for many of the more useful types based linting rules, but it also is not strictly necessary either.

Not totally sure, but we might need to disable/enable certain rules based on whether the `tsconfig.json` file was found, along with a warning perhaps?

**Comment by Divlo** [comment]
Also is this issue #1283 is related to this one ?

**Comment by feross** [comment]
Hmm, I'm not familiar with `tsconfig.json`. My first impression is that adding a config file doesn't seem in the spirit of `standard`. Ideally we still support the workflow of `npx standard` and you automatically get useful output.

What sorts of rules are enabled by adding `tsconfig.json` and how important is it?

**Comment by toddbluhm** [comment]
## TLDR; 

A `tsconfig.json` or `tsconfig.eslint.json` file is required if we wish to take full advantage of the `typescript-eslint` parser with type-aware rules. The current `eslint-config-standard-with-typescript` uses some type-aware rules, but I am not sure which ones are type-aware.  We should probably document that in `eslint-config-standard-with-typescript` at some point.

## tsconfig.json

A `tsconfig.json` file is a project file that tells the typescript compiler what files to compile, what version of the compiler to use, and a few other compiler options. A `tsconfig.json` is found in almost every project (some extremely simple projects may not need one, but any sort of complex project would require one). Here are the official docs on [`tsconfig.json`](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)

## typescript-eslint

The [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint) parser, uses the `tsconfig.json` file to figure out which files are part of the typescript project and some compiler configuration information. In order to get the most accurate type of information for type-aware eslint rules, all project files must be known and parsed/compiled before the rules can be applied. Here is a quick little blurb from the `typescript-eslint` project about it [typescript-eslint: TYPED_LINTING](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/TYPED_LINTING.md#i-get-errors-telling-me-the-file-must-be-included-in-at-least-one-of-the-projects-provided). `eslint-config-standard-with-typescript` site also links to this `typescript-eslint` doc [typescript-eslint: parseroptions.project](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/parser/README.md#parseroptionsproject)

The naming of the `tsconfig.json` file can change as well. Some people do `tsconfig.production.json`, `tsconfig.staging.json`, etc... but that is probably out of the scope of this discussion.

## tsconfig.eslint.json

What about the `tsconfig.eslint.json` file?! This is where things get interesting. When building a typescript project, I really don't want to compile my test case files, etc... into my typescript project, but I still want to lint those files. So I have a `tsconfig.eslint.json` that is essentially a `tsconfig.json` file, but it includes all typescript files in my project. This file is not picked up by the typescript compiler but is picked up by `typescript-eslint`. So maybe we should be focusing on using the `tsconfig.eslint.json` instead of the `tsconfig.json` file?

For `ts-standard` I look for `tsconfig.eslint.json` first, then check for `tsconfig.json` in the root directory. The presence of one of those files might be a good way of detecting whether to enable typescript support in `standard` automatically?

## no tsconfig.json file

We could always go with non-type-aware rules and then we would not have to worry about needing a project file at all (`tsconfig.json`/ `tsconfig.eslint.json`), however, we would lose the significant benefits of typescript with its compiler/AST.

## opinion

I think the most sensible option is to default to assuming the `tsconfig.json`/`tsconfig.eslint.json` file is in the root project directory. We could allow the user to set which project file is used via a CLI flag or `package.json` setting.

If we want to go the route of always requiring the `tsconfig.json`/`tsconfig.eslint.json` file be in the root directory and allow no other configuration, then that might work as well. Just say we are setting a `standard`.

I would not recommend removing type-aware rules just to work around setting/getting the `tsconfig.json`/`tsconfig.eslint.json` project file. A major point in using typescript is the type-information and the code improvements and context that brings with it.

**Comment by Divlo** [comment]
We should close this issue since `ts-standard` is now a thing.

**Comment by feross** [comment]
This issue has morphed from discussion of adding `eslint-config-standard-with-typescript` to the @standard org (which is done) to discussion of how to build in TypeScript support directly into `standard`. I suppose that `ts-standard` removes the pressing need for this by providing a nice happy path for TS users, so we could close this and end discussion of adding TS support directly into `standard` for now if we'd like.
